app/src/androidTest/java/org/mozilla/fenix/ui/BadServerResponseTest.kt

```diff
@@ -0,0 +1,88 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.ui
+
+import okhttp3.mockwebserver.MockWebServer
+import org.junit.After
+import org.junit.Test
+import org.mozilla.fenix.helpers.HomeActivityTestRule
+import org.mozilla.fenix.helpers.MockWebServerHelper
+import org.mozilla.fenix.ui.robots.homeScreen
+
+class BadServerResponseTest {
+    private val activityTestRule = HomeActivityTestRule()
+    private var mockWebServer = MockWebServer()
+
+    @After
+    fun tearDown() {
+        mockWebServer.shutdown()
+    }
+
+    @Test
+    fun serverResponse400StartupTest() {
+        mockWebServer = MockWebServerHelper.createErrorResponseMockWebServer(400)
+        mockWebServer.start()
```

Comment:
I don't think this is working to intercept requests. I added some code inside `HomeActivity.onCreate` to make a request:
```kotlin
        GlobalScope.async(Dispatchers.IO) {
            val req = Request("https://example.com")
            val res = components.core.client.fetch(req)
            android.util.Log.e("lol", "$res")
        }
```

And when I ran this test, the logged response status code was always 200.

I think we may need to actually inject the URLs returned by `mockWebServer` into the production code to get mocked network request values (or to mock the fetch client in components). I'm not sure how hard this would be but if it gets complex, maybe it's not worth writing such tests.

Reply:
Haven't tested but maybe we could:
- wrap the `GeckoViewFetchClient` from [Core.client](https://github.com/mozilla-mobile/fenix/blob/e07f4ac590f85d4bc1209cff1325315aeb4c0ba2/app/src/main/java/org/mozilla/fenix/components/Core.kt#L151) in an ~`AppFetchClient`, with the same contract, it just delegating the calls further to
    - in production the current `GeckoViewFetchClient`.
    - in tests a Fake for which we can easily modify any part of it's behavior (like the base url).
   
 I think Nimbus may be using it's own http client so maybe we could also   
 - declare in `androidTest` a similar [`createNimbus` method](https://github.com/mozilla-mobile/fenix/blob/e07f4ac590f85d4bc1209cff1325315aeb4c0ba2/app/src/main/java/org/mozilla/fenix/experiments/NimbusSetup.kt#L21) that would allow for the same as above:
    - in tests to return `NimbusApi`Fake or even the real one but initialized with the base url that we want. 
    
Going with the above would ensure all modifications are in Fenix with no change needed in any dependency, easily visible and easy to control.

Thoughts?

Reply:
> wrap the GeckoViewFetchClient from Core.client in an ~AppFetchClient, with the same contract, it just delegating the calls further to
    in production the current GeckoViewFetchClient.
    in tests a Fake for which we can easily modify any part of it's behavior (like the base url).

This sounds like it'd work but, if possible, I'd prefer to substitute the implementation of GeckoViewFetchClient with a Fake rather than wrap it to enable the Fake (it seems like one less layer of indirection and it's avoiding test branches from possibly entering production).

In Robolectric integration tests, we're able to do this by providing a different `Application` with its own `TestComponents`:
https://github.com/mozilla-mobile/fenix/blob/85c6ef2cf7f3520b8fa1e410146d364f6feb3741/app/src/test/java/org/mozilla/fenix/helpers/FenixRobolectricTestApplication.kt#L23

And that extends `Components` to replace a few. For example, we mock the background services for all tests (unless they undo this behavior):
https://github.com/mozilla-mobile/fenix/blob/85c6ef2cf7f3520b8fa1e410146d364f6feb3741/app/src/test/java/org/mozilla/fenix/components/TestComponents.kt#L14-L16

I'm not sure if there's a similar implementation opportunity for UI tests though.

> declare in androidTest a similar createNimbus method that would allow for the same as above:
    in tests to return NimbusApiFake or even the real one but initialized with the base url that we want.

This seems reasonable. Ideally, we could pass in an HttpClient like we do above (for consistency reasons) but if that's not possible (e.g. they're using a Rust HttpClient and not the android components `Client` interface) then replacing the implementation at a higher level seems reasonable.

This implementation could also be a follow-up: we do most of the app with the `Client` implementation Fake above and then implement the Nimbus thing separately.

Unrelated, if nimbus uses its own Http client, I wonder if the Tor browser correctly intercepts and proxies the requests it makes. It might be good to double-check this.

Reply:
Thank you!
Will look into the needed modifications to support overriding components in ui tests.
I like the idea of having a separate runner/application with overriden components!

Regarding Nimbus' http client there is a comment in code about a different one.
https://github.com/mozilla-mobile/fenix/blob/993428cd0f994ccfa1b6f9fd1c724e6ad0ff45ec/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L320-L321

Don't know how tor packages firefox. In the case of Nimbus I would think they are not providing the `NIMBUS_ENDPOINT` needed at compile time to easily disable the experiments.
Tried https://support.torproject.org/get-in-touch/#irc-help to just start a discussion but could not get to connect to their IRC. Will try later.

Reply:
Was planning to work on this ^^ in between other tickets but I won't have time in the next few days.
Filed https://github.com/mozilla-mobile/fenix/issues/18385 if somebody else wants to help.